pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/EmptyResponseUtils.java

```diff
@@ -0,0 +1,206 @@
+    List<Pair<AggregationFunction, FilterContext>> filteredAggregationFunctions =
+        queryContext.getFilteredAggregationFunctions();
+    assert filteredAggregationFunctions != null;
+    int numAggregations = filteredAggregationFunctions.size();
+    String[] columnNames = new String[numAggregations];
+    ColumnDataType[] columnDataTypes = new ColumnDataType[numAggregations];
+    Object[] row = new Object[numAggregations];
+    for (int i = 0; i < numAggregations; i++) {
+      Pair<AggregationFunction, FilterContext> pair = filteredAggregationFunctions.get(i);
+      AggregationFunction aggregationFunction = pair.getLeft();
+      columnNames[i] = AggregationFunctionUtils.getResultColumnName(aggregationFunction, pair.getRight());
+      columnDataTypes[i] = aggregationFunction.getFinalResultColumnType();
+      row[i] = aggregationFunction.extractFinalResult(
+          aggregationFunction.extractAggregationResult(aggregationFunction.createAggregationResultHolder()));
+    }
+    return new ResultTable(new DataSchema(columnNames, columnDataTypes), List.<Object[]>of(row));
+  }
+
+  private static ResultTable buildEmptyGroupByResultTable(QueryContext queryContext) {
+    List<Pair<AggregationFunction, FilterContext>> filteredAggregationFunctions =
+        queryContext.getFilteredAggregationFunctions();
+    List<ExpressionContext> groupByExpressions = queryContext.getGroupByExpressions();
+    assert filteredAggregationFunctions != null && groupByExpressions != null;
+    int numColumns = groupByExpressions.size() + filteredAggregationFunctions.size();
+    String[] columnNames = new String[numColumns];
+    ColumnDataType[] columnDataTypes = new ColumnDataType[numColumns];
+    int index = 0;
+    for (ExpressionContext groupByExpression : groupByExpressions) {
+      columnNames[index] = groupByExpression.toString();
+      // Use STRING column data type as default for group-by expressions
+      columnDataTypes[index] = ColumnDataType.STRING;
+      index++;
+    }
+    for (Pair<AggregationFunction, FilterContext> pair : filteredAggregationFunctions) {
+      AggregationFunction aggregationFunction = pair.getLeft();
+      columnNames[index] = AggregationFunctionUtils.getResultColumnName(aggregationFunction, pair.getRight());
+      columnDataTypes[index] = aggregationFunction.getFinalResultColumnType();
+      index++;
+    }
+    return new ResultTable(new DataSchema(columnNames, columnDataTypes), List.of());
+  }
+
+  /**
+   * Tries to fill an {@link DataSchema} when no row has been returned.
+   *
+   * Response data schema can be inaccurate (all columns set to STRING) when all segments are pruned on broker or
+   * server.
+   *
+   * Priority is:
+   * - Types from multi-stage engine validation for the given query (if allowed).
+   * - Types from schema for the given table (only applicable to identifiers).
+   * - Types from single-stage engine response (no action).
```

Comment:
nit: here is where markdown style shines, but here were are using old javadoc comments